text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+
+![Traffic reduction](img/0156_OTEL%20-%20InternetTrafficReduction.png)
+
+> Note 1: A fallback mechanism can be used to handle the case where the new protocol is not supported by the target.
+> More on this mechanism in this [section](#protocol-extension-and-fallback-mechanism).
+
+#### Phase 2
+
+Phase 2 aims to extend the support of Apache Arrow end-to-end and more specifically inside the collector to better
+support the following scenarios: cost efficiency, advanced data processing, data minimization. New receivers, processors,
+and exporters supporting Apache Arrow natively will be developed. A bidirectional adaptation layer OTLP / OTLP Arrow
+will be developed within the collector to continue supporting the existing ecosystem. The following diagram is an
+overview of a collector supporting both OTLP and an end-to-end OTLP Arrow pipeline.
+
+![OTEL Arrow Collector](img/0156_OTEL%20-%20OTELArrowCollector.png)
+
+Implementing an end-to-end column-oriented pipeline will provide many benefits such as:
+
+- **Accelerate stream processing**,
+- **Reduce CPU and Memory usage**,
+- **Improve compression ratio end-to-end**,
+- **Access to the Apache Arrow ecosystem** (query engine, parquet support, ...).
+
+## Protocol Details
+
+The protocol specifications are composed of two parts. The first section describes the new gRPC service supporting
+column-oriented telemetry data. The second section presents the mapping between the OTLP entities and their Apache
+Arrow counterpart.
+
+### ArrowStreamService
+
+OTLP Arrow defines the columnar encoding of telemetry data and the gRPC-based protocol used to exchange data between
+the client and the server. OTLP Arrow is a bi-directional stream oriented protocol leveraging Apache Arrow for the
+encoding of the telemetry data.
+
+OTLP and OTLP Arrow protocols can be used together and can use the same TCP port. To do so, in addition to the 3
+existing
+services (`MetricsService`, `LogsService` and `TraceService`), we introduce the service `ArrowStreamService`
+(see [this protobuf specification](#appendix-a---protocol-buffer-definitions) for more details) exposing a single API endpoint named `ArrowStream`.
+This endpoint is based on a bidirectional streaming protocol. The ingress side is a `BatchArrowRecords` stream encoding a
+batch of Apache Arrow buffers (more specifically [Arrow IPC format](#arrow-ipc-format)). The egress
+side is a `BatchStatus` stream reporting asynchronously the status of each `BatchArrowRecords` previously sent.
+
+After establishing the underlying transport the client starts sending telemetry data using the `ArrowStream` request.
+The
+client continuously sends `BatchArrowRecords`'s messages over the opened stream to the server and expects to receive
+continuously
+`BatchStatus`'s messages from the server as illustrated by the following sequence diagram:
+
+![Sequence diagram](img/0156_OTEL%20-%20ProtocolSeqDiagram.png)
```

Comment:
I wrote a diagram to explain the approach taken:
https://github.com/open-telemetry/experimental-arrow-collector/blob/main/exporter/otlpexporter/arrow-design.png

The Arrow exporter has a `num_streams` limit that controls how many arrow-record producers are maintained concurrently; each stream sends as fast as it can by dequeuing one of the outstanding work items, but the `NumConsumers` configuration applied by `exporterhelper` controls how many records can be outstanding at once. Each stream has a concurrent receiver awaiting the status of each outstanding request, at which point the `exporterhelper` call returns.

The gRPC send/recv buffers will also be effective at maintaining backpressure.